lib/Dialects/LinalgTransform/Transforms/TransformInterpreter.cpp

```diff
@@ -514,18 +641,13 @@ static void noOpRewriter(ArrayRef<PDLValue> args, ArrayAttr constantParams,
 #endif
 }
 
-/// Returns the operations that `transformOp` should be run on. The
-/// `transformOp` is expected to have the TargetableTransformOpTrait. Its target
-/// may be either an SSA value that corresponds to a list of ops produced by
-/// other transforms and stored in `operations` or a list of ops that match the
-/// PDL pattern specified by its name.
 template <typename OpTy>
-static FailureOr<SmallVector<LinalgOp>>
-findTransformTarget(OpTy transformOp, ModuleOp module,
-                    const DenseMap<Value, SmallVector<LinalgOp>> &operations) {
-  if (Value target = transformOp.target())
-    return operations.lookup(target);
+using has_target_matcher_t = decltype(std::declval<OpTy &>().targetMatcher());
```

Comment:
Should this be `target_matcher_t` ?
The `has_` prefix suggests a boolean type but I don't see it evaluating to a bool ..

Reply:
The `_t` suffix combined with the STL naming scheme indicates that name corresponds to a type. This is the normal idiom for `llvm::is_detected` (https://llvm.org/doxygen/namespacellvm.html#a5222d74738fef771d06ccec12e74cf97), which is used just below.

Reply:
I would have expected `has_xxx_t` to correspond to either `true_type` or `false_type` from type_traits (see e.g.  https://en.cppreference.com/w/cpp/header/type_traits).

The example you post says: `has_copy_assign_t = decltype(std::declval<T&>() = std::declval<const T&>()); `.
I would think that this evals to true_type / false_type, though didn't check and not 100% sure as I don't know what `std::declval<T&>() = std::declval<const T&>()` evaluates to.

Reply:
No, that does not evaluate to `std::true/false_type`, unless the copy-assignment of T returns an instance of that type. (Technically speaking, it does not evaluate to anything at all since decltype is an unevaluated context.) `std::declval<T&>() = std::declval<const T&>() ` is equivalent to `std::declval<T &>().operator =(std::declval<const T&>())`, so this is an expression that calls the copy-assignment operator. Putting this in `decltype` specifies the type that the expression would have had when evaluated. In this case, it is the return type of `T::operator =(const T&)`. In most cases, it is just `T &` to allow for chained assignments, but that is actually irrelevant. The important part is the expression inside `decltype` being well-formed, i.e. the copy-assignment operator existing on `T` in the example case. The type alias defined with this `decltype` is used a s template argument for `llvm::is_detected` so if the expression inside `decltype` is ill-formed, template argument substitution fails in the specialized version (that aliases `is_detected` to `true_type`), the non-specialized version (that aliases `is_detected` to `false_type`) can be used since substitution failure is not an error.

Reply:
I see, thanks for spelling it out.
So the true/false interpretation is actually in TM^2P land: error or not and this gets resolved by SFINAE .. gah